db/db_impl.h

```diff
@@ -1213,6 +1219,10 @@ class DBImpl : public DB {
   std::unordered_map<uint64_t, uint64_t> prepared_section_completed_;
   std::mutex prep_heap_mutex_;
 
+  // Callback for compaction to check if a key is visible to a snapshot.
+  // REQUIRES: mutex held
```

Comment:
why?

Reply:
smart pointers are not thread-safe, and I'm not making it constant since I set it during WritePrepareTxnDB::init() which is after DBImpl open.

Reply:
Does it any overhead? Does compaction already holds the mutex_?

Reply:
We only set the field once. Compaction hold mutex on-and-off. Just make sure the pointer is fetched while compaction is holding mutex.